### ISSUE TITLE
TransformedDistribution's log_prob gradient is inconsistent w.r.t. cache_size

### ISSUE THREAD

**Issue by fehiepsi** [pytorch/pytorch#59822]
## 🐛 Bug

Currently, when we draw a non-reparameterized sample from TransformedDistribution and evaluate its log probability, we get different results when changing the cache_size of the transform.

## To Reproduce

```python
import torch
import torch.distributions as dist
torch.manual_seed(0)
power = torch.tensor(2., requires_grad=True)
d = dist.TransformedDistribution(
    dist.LogNormal(0., 1.), dist.transforms.PowerTransform(power, cache_size=1))
z = d.sample()
d.log_prob(z).backward()
print(power.grad)  # return -0.500
torch.manual_seed(0)
power = torch.tensor(2., requires_grad=True)
d = dist.TransformedDistribution(
    dist.LogNormal(0., 1.), dist.transforms.PowerTransform(power, cache_size=0))
z = d.sample()
d.log_prob(z).backward()
print(power.grad)  # return 0.6873
```

## Expected behavior

The result should be agnostic w.r.t. cache_size. It seems to me that the result with `cache_size=0` is correct.

## Environment

 - PyTorch Version: 1.8.1
 - OS: Linux
 - How you installed PyTorch: conda
 - Python version: 3.8

## Additional context

The discussion here is relevant: https://github.com/pyro-ppl/numpyro/pull/1044#discussion_r642103731

cc @martinjankowiak @fritzo
